pkg/server/home_workspaces.go

```diff
@@ -327,145 +181,182 @@ func (h *homeWorkspaceHandler) ServeHTTP(rw http.ResponseWriter, req *http.Reque
-			h.apiHandler.ServeHTTP(rw, req)
-			return
+		// nothing to check permission-wise. If there is a very improbable hash conflict, the will get a 403
+		// through normal authorization. We don't give access to legacy home workspaces to new users here.
+
+		homeWorkspace := &tenancyv1beta1.Workspace{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: this.Name,
+			},
+			Spec: tenancyv1beta1.WorkspaceSpec{
+				Type: this.Spec.Type,
+			},
+			Status: tenancyv1beta1.WorkspaceStatus{
+				URL:          this.Status.URL,
+				Phase:        this.Status.Phase,
+				Conditions:   this.Status.Conditions,
+				Initializers: this.Status.Initializers,
+			},
 		}
+		responsewriters.WriteObjectNegotiated(homeWorkspaceCodecs, negotiation.DefaultEndpointRestrictions, tenancyv1beta1.SchemeGroupVersion, rw, req, http.StatusOK, homeWorkspace)
+		return
+	}
 
-		var needsAutomaticCreation bool
-		needsAutomaticCreation, workspaceType = h.needsAutomaticCreation(lcluster.Name)
-		logger.V(4).Info("not a ~ request", "needsAutoCreate", needsAutomaticCreation, "verb", requestInfo.Verb, "url", req.URL)
-		if !needsAutomaticCreation {
-			h.apiHandler.ServeHTTP(rw, req)
+	// here we have not found a legacy home workspace. We have to check whether a new-style workspace exists.
+
+	homeClusterName := homeClusterName(effectiveUser.GetName())
+	var this *tenancyv1alpha1.ThisWorkspace
+	objs, err := h.thisWorkspaceIndexer.ByIndex(indexers.ByLogicalCluster, homeClusterName.String())
+	if err != nil {
+		responsewriters.InternalError(rw, req, err)
+		return
+	} else if len(objs) > 1 {
+		logger.Info("multiple ThisWorkspaces found for the same logical cluster", "logicalCluster", legacyHomeParent.Join(legacyHomeName).String())
+		responsewriters.InternalError(rw, req, fmt.Errorf("unexpectedly multiple home workspaces found"))
+		return
+	} else if len(objs) == 0 {
+		// check permissions first
+		attr := authorizer.AttributesRecord{
+			User:            effectiveUser,
+			Verb:            "create",
+			APIGroup:        tenancyv1alpha1.SchemeGroupVersion.Group,
+			Resource:        "workspaces",
+			Name:            "~",
+			ResourceRequest: true,
+		}
+		decision, _, err := h.authz.Authorize(ctx, attr)
+		if err != nil {
+			logger.WithValues("cluster", homeClusterName, "user", effectiveUser.GetName()).Error(err, "error authorizing request")
+			responsewriters.Forbidden(ctx, attr, rw, req, authorization.WorkspaceAccessNotPermittedReason, homeWorkspaceCodecs)
+			return
+		}
+		if decision != authorizer.DecisionAllow {
+			responsewriters.Forbidden(ctx, attr, rw, req, authorization.WorkspaceAccessNotPermittedReason, homeWorkspaceCodecs)
 			return
 		}
 
-		foundLocally, retryAfterSeconds, err := h.searchForWorkspaceAndRBACInLocalInformers(lcluster.Name, workspaceType == HomeClusterWorkspaceType, effectiveUser.GetName())
-		if err != nil {
+		logger.Info("Creating home ThisWorkspace", "cluster", homeClusterName.String(), "user", effectiveUser.GetName())
```

Comment:
which controller creates the corresponding `ClusterWorkspace` ?

Reply:
None. There is no ClusterWorkspace for home workspaces. They are parent-less.